pkg/geo/geogfn/segmentize.go

```diff
@@ -0,0 +1,161 @@
+	case *geom.Polygon:
+		segPolygon := geom.NewPolygon(geom.XY).SetSRID(geometry.SRID())
+		for loopIdx := 0; loopIdx < geometry.NumLinearRings(); loopIdx++ {
+			l, err := segmentizeGeom(geometry.LinearRing(loopIdx), segmentMaxLength)
+			if err != nil {
+				return nil, err
+			}
+			err = segPolygon.Push(l.(*geom.LinearRing))
+			if err != nil {
+				return nil, err
+			}
+		}
+		return segPolygon, nil
+	case *geom.MultiPolygon:
+		segMultiPolygon := geom.NewMultiPolygon(geom.XY).SetSRID(geometry.SRID())
+		for polygonIdx := 0; polygonIdx < geometry.NumPolygons(); polygonIdx++ {
+			p, err := segmentizeGeom(geometry.Polygon(polygonIdx), segmentMaxLength)
+			if err != nil {
+				return nil, err
+			}
+			err = segMultiPolygon.Push(p.(*geom.Polygon))
+			if err != nil {
+				return nil, err
+			}
+		}
+		return segMultiPolygon, nil
+	case *geom.GeometryCollection:
+		segGeomCollection := geom.NewGeometryCollection().SetSRID(geometry.SRID())
+		for geoIdx := 0; geoIdx < geometry.NumGeoms(); geoIdx++ {
+			g, err := segmentizeGeom(geometry.Geom(geoIdx), segmentMaxLength)
+			if err != nil {
+				return nil, err
+			}
+			err = segGeomCollection.Push(g)
+			if err != nil {
+				return nil, err
+			}
+		}
+		return segGeomCollection, nil
+	}
+	return nil, errors.Newf("unknown type: %T", geometry)
+}
+
+// segmentizeCoords inserts multiple points between given two-coordinate and
+// return resultant points as []float64. Such that distance between any two
+// points is less than given maximum segment's length, the total number of
+// segments is the power of 2, and all the segments are of the same length.
+// NOTE: List of points does not consist of end point.
+func segmentizeCoords(a geom.Coord, b geom.Coord, maxSegmentLength float64) []float64 {
+	// Converted geom.Coord into s2.Point so we can segmentize the coordinates.
+	point1 := s2.PointFromLatLng(s2.LatLngFromDegrees(a.Y(), a.X()))
+	point2 := s2.PointFromLatLng(s2.LatLngFromDegrees(b.Y(), b.X()))
+
+	var segmentizedPoints []s2.Point
+	distanceBetweenPoints := s2.ChordAngleBetweenPoints(point1, point2).Angle().Radians()
+	if maxSegmentLength <= distanceBetweenPoints {
+		// This calculation is to determine the total number of segment between given
+		// 2 coordinates. For that fraction by segment must be less than or equal to
+		// the fraction of max segment length to distance between point, since the
+		// total number of segment must be power of 2 therefore we can write as
+		// 1 / (2^n)[numberOfSegmentToCreate] <= segmentMaxLength / distanceBetweenPoints < 1 / (2^(n-1))
+		// (2^n)[numberOfSegmentToCreate] >= distanceBetweenPoints / segmentMaxLength > 2^(n-1)
+		// therefore n = ceil(log2(segmentMaxLength/distanceBetweenPoints)). Hence
+		// numberOfSegmentToCreate = 2^(ceil(log2(segmentMaxLength/distanceBetweenPoints))).
+		numberOfSegmentToCreate := int(math.Pow(2, math.Ceil(math.Log2(distanceBetweenPoints/maxSegmentLength))))
```

Comment:
ok, this math is a little complicated. i think doing some inversing makes this cleaner:
```
numPoints := math.Ceil(maxSegmentLength / distanceBetweenPoints)
```
is there something wrong with doing it this way around?

Reply:
I think there may be a problem with directly writing it that way,
for example,
```
postgis_db=# select st_distance('POINT(0 0)'::geography, 'POINT(1 1)'::geography, false);
   st_distance   
-----------------
 157249.59776851
(1 row)

select st_astext(ST_Segmentize('LINESTRING(0 0, 1 1'::geography, 78624.798888))

postgis_db=# select st_astext(ST_Segmentize('LINESTRING(0 0, 1 1)'::geography, 78624.798888));
                        st_astext                        
---------------------------------------------------------
 LINESTRING(0 0,0.499961919922622 0.500019038226106,1 1)

postgis_db=# select st_astext(ST_Segmentize('LINESTRING(0 0, 1 1)'::geography, 78623.798888));
                                                            st_astext                                                            
---------------------------------------------------------------------------------------------------------------------------------
 LINESTRING(0 0,0.249976200223564 0.250011898653433,0.499961919922622 0.500019038226106,0.749966679297834 0.750016659002919,1 1)
```

for the case, one ratio will be 1.99999 
and the total number of segments was 2, and math.Ceil(distanceBetweenPoints/maxSegmentLength)
will as give 2.
and for case 2 ratio was 2.0002
and the total number of segments was 4, and math.Ceil(distanceBetweenPoints/maxSegmentLength)
will as give 3.

Reply:
Looks like I misunderstood what this does. Thanks for setting me right.